text/0004-type-deduction.md

```diff
@@ -0,0 +1,214 @@
+It is in question what to return here best dependend on the operator in use. One might solves this elegantly by handling
+this case the same as for optional positional arguments with dependencies between variables.
+(Note: One might also think about this not as a graph problem, but as an CSP.)
+
+## Merging of deductions.
+Any variable can occur multiple times within the block. In each position the variable can have different possible Types.
+```shell Example
+config | where size < 1kb * $a; kill $a
+$a -> SyntaxShape::Int
+```
+In the first position, $a might be SyntaxShape::Int or SyntaxShape::Number. In the second it is of SyntaxShape::Int.
+The result set of possible types for a variable can be computed as the set intersection between the already deduced types and the new deduced types. If the intersection is empty, it means that the variable has usages with different and non compatible types. An error is returned.
+
+```rust
+fn checked_insert(existing_deductions, new_deductions) -> Result<set_of_deductions, ShellError>{
+    if existing_deductions == None return new_deductions
+    else return set_intersection(existing_deductions, new_deductions)?
+}
+
+existing_deductions = checked_insert(existing_deductions, new_deductions);
+```
+
+Special cases arise if one (or both) sets contains SyntaxShape::Any. SyntaxShape::Any is a placeholder for any possible type. The above pseudo code has to be changed to: 
+```rust
+fn checked_insert(existing_deductions, new_deductions) -> Result<set_of_deductions, ShellError>{
+    if existing_deductions == None return new_deductions
+    else match (has_any(existing_deductions), has_any(new_deductions)){
+        (true, true) -> set_union_including_any(existing_deductions, new_deductions)
+        (true, false) -> set_union_excluding_any(new_deductions, set_intersection(existing_deductions, new_deductions))
+        (false, true) -> set_union_excluding_any(existing_deductions, set_intersection(existing_deductions, new_deductions))
+        (false, false) -> set_intersection(existing_deductions, new_deductions)?
+    }
+}
+
+existing_deductions = checked_insert(existing_deductions, new_deductions);
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+A start may be: https://github.com/nushell/nushell/pull/2486
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+None I can think of.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+# Prior art
+
+[prior-art]: #prior-art
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+## Deducing of result type of column paths
+Currently it is not possible to infer the shape of column paths pointing to a table given by a prior command in the pipeline. 
+```shell $size nor $name are deducable
+ls | where size == $size | get name | where $it == $name
```

Comment:
We probably can't reasonably do this, yet, until we have input/output types in place